docs/architecture/adr-110-remote-mempool.md

```diff
@@ -0,0 +1,364 @@
+  clusters relatively straightforward.
+
+An alternative here would be to implement a RESTful HTTP/3 API (using QUIC), but
+HTTP/3 support in reverse proxies is still in its early stages. Such a protocol
+can be considered in a future iteration of the interface if it provides
+substantial performance/latency benefits.
+
+The following gRPC API is proposed.
+
+```protobuf
+syntax = "proto3";
+package tendermint.services.remote_mempool.v1;
+
+// RemoteMempoolService is implemented by a remote mempool instance. A CometBFT
+// validator makes use of a client generated from this interface definition to
+// interact with the remote mempool.
+//
+// It is up to the implementor of the remote mempool to define transaction
+// submission mechanisms/interfaces.
+service RemoteMempoolService {
+    // Fetch allows a CometBFT validator to obtain the next batch of
+    // transactions to be included during PrepareProposal when the current node
+    // is a proposer.
+    rpc Fetch(FetchRequest) returns (FetchResponse);
+
+    // Reap takes a set of transaction keys and removes the corresponding
+    // transactions from the mempool.
+    //
+    // This will only be called after the block is committed by consensus.
+    rpc Reap(ReapRequest) returns (ReapResponse);
+
+    // TxsAvailable streams notifications back to the client that new
+    // transactions are available in the mempool.
+    //
+    // The mempool is expected to stream a TxsAvailableResponse as it has
+    // transactions available. A response, however, is only expected in two
+    // cases:
+    //
+    // 1. Once for each newly encountered height (as supplied to the mempool via
+    //    the Reap request).
+    //
+    // 2. Upon initiation of the TxsAvailable call. This caters for instances
+    //    where the consensus engine may have failed and restarted.
+    //
+    // The consensus engine will only ever make one call to TxsAvailable and
+    // will attempt to keep the connection alive.
+    rpc TxsAvailable(TxsAvailableRequest) returns (stream TxsAvailableResponse);
+}
+
+message FetchRequest {
+    // Fetch as many transactions as possible that cumulatively take up at most
+    // this number of bytes. Setting this value to -1 implies no limit (this
+    // implies that the remote mempool is fully in control of this value).
+    int64 max_bytes = 1;
+}
+
+message FetchResponse {
+    // Transactions to be included in the proposal.
+    repeated bytes txs = 1;
+}
+
+message ReapRequest {
+    // The current height of the chain.
+    uint64 height = 1;
+
+    // A list of IDs of transactions to be removed from the mempool. At present
+    // this is a list of SHA256 hashes of the transactions.
+    repeated bytes tx_ids = 2;
+}
+
+message ReapResponse {}
+
+message TxsAvailableRequest {}
+
+// TxsAvailableResponse is to be sent once upon initiation of the TxsAvailable
+// request, as well as once for each new height, iff new transactions are
```

Comment:
Should we define the behavior for cases where remote mempool sends `TxsAvailableResponse` more than once per height?